kuzzle/kuzzle.go

```diff
@@ -87,36 +132,79 @@ func NewKuzzle(c protocol.Protocol, options types.Options) (*Kuzzle, error) {
 	return k, nil
 }
 
-// Connect connects to a Kuzzle instance using the provided host and port.
+// Connect connects to a Kuzzle instance.
 func (k *Kuzzle) Connect() error {
-	wasConnected, err := k.socket.Connect()
-	if err == nil {
-		if k.lastUrl != k.socket.Host() {
-			k.wasConnected = false
-			k.lastUrl = k.socket.Host()
+	if k.autoQueue {
+		k.queuing = true
+	}
+
+	wasConnected, err := k.protocol.Connect()
+
+	if err != nil {
+		if k.autoReconnect && !k.retryingToConnect && !k.stopRetryingToConnect && k.nbConnectionRetried < MAX_CONNECT_RETRY {
+			k.retryingToConnect = true
+			time.Sleep(k.reconnectionDelay * time.Millisecond)
+			k.retryingToConnect = false
+			k.nbConnectionRetried++
+			return k.Connect()
 		}
 
-		if wasConnected {
-			if k.jwt != "" {
-				go func() {
-					res, err := k.Auth.CheckToken(k.jwt)
-
-					if err != nil {
-						k.jwt = ""
-						k.socket.EmitEvent(event.TokenExpired, nil)
-						return
-					}
-
-					if !res.Valid {
-						k.jwt = ""
-						k.socket.EmitEvent(event.TokenExpired, nil)
-					}
-				}()
-			}
+		return types.NewError(err.Error())
+	}
+
+	k.nbConnectionRetried = 0
+	k.stopRetryingToConnect = false
+	k.queuing = false
+
+	if k.lastUrl != k.protocol.Host() {
+		k.wasConnected = false
+		k.lastUrl = k.protocol.Host()
+	}
+
+	if wasConnected {
+		if k.jwt != "" {
+			go func() {
+				res, err := k.Auth.CheckToken(k.jwt)
+
+				if err != nil {
+					k.jwt = ""
+					k.protocol.EmitEvent(event.TokenExpired, nil)
+					return
+				}
+
+				if !res.Valid {
+					k.jwt = ""
+					k.protocol.EmitEvent(event.TokenExpired, nil)
+				}
+			}()
 		}
-		return nil
+
+		k.EmitEvent(event.Reconnected, nil)
```

Comment:
The `reconnected` event must be emitted _after_ the JWT has been checked and, eventually, invalidated, to make sure that one cannot send an API request with an obsolete JWT (see the JS implementation: https://github.com/kuzzleio/sdk-javascript/blob/master/src/Kuzzle.js#L277 )